File: apps/android/app/src/main/java/app/ollo/messenger/data/AuthRepository.kt
package app.ollo.messenger.data

import app.ollo.crypto.SessionController
import app.ollo.crypto.SessionSecrets
import app.ollo.network.OlloApi
import org.json.JSONObject

class AuthRepository(
    private val api: OlloApi,
    private val sessions: SessionController? = null,
) {
    fun requestOtp(phone: String): Pair<String, String?> {
        val raw = api.requestOtp(phone)
        val json = JSONObject(raw)
        val dev = if (json.has("dev_otp") && !json.isNull("dev_otp")) json.getString("dev_otp") else null
        return json.getString("challenge_id") to dev
    }

    fun verify(challengeId: String, otp: String, deviceJson: String): Session {
        val body = JSONObject()
            .put("challenge_id", challengeId)
            .put("otp", otp)
            .put("device", JSONObject(deviceJson))
            .toString()
        val raw = api.post("/v1/auth/verify-otp", body, auth = false)
        val json = JSONObject(raw)
        val user = json.getJSONObject("user")
        val session = Session(
            userId = user.getString("id"),
            username = if (user.isNull("username")) null else user.getString("username"),
            deviceId = json.getString("device_id"),
            access = json.getString("access_token"),
            refresh = json.getString("refresh_token"),
        )
        sessions?.save(SessionSecrets(session.userId, session.deviceId, session.access, session.refresh))
        return session
    }

    fun logout() {
        sessions?.wipe()
    }

    data class Session(
        val userId: String,
        val username: String?,
        val deviceId: String,
        val access: String,
        val refresh: String,
    )

    companion object {
        /**
         * Authenticated client: 401 retries once via refresh, then wipes the
         * protocol store. [deviceJson] is still produced only by a bound engine.
         */
        fun connected(baseUrl: String, sessions: SessionController): AuthRepository {
            lateinit var api: OlloApi
            api = OlloApi(
                baseUrl,
                token = { sessions.access() },
                refresh = refresh@{
                    val token = sessions.refresh() ?: return@refresh false
                    val raw = try {
                        api.refreshSession(token)
                    } catch (_: Throwable) {
                        return@refresh false
                    }
                    val json = JSONObject(raw)
                    val access = json.optString("access_token")
                    val refreshTok = json.optString("refresh_token")
                    if (access.isEmpty() || refreshTok.isEmpty()) return@refresh false
                    sessions.applyRefresh(access, refreshTok)
                },
                onWipe = { sessions.wipe() },
            )
            return AuthRepository(api, sessions)
        }
    }
}

File: apps/ios/Sources/OlloNetwork/AuthRepository.swift
import Foundation
import OlloCrypto

/// Twin of Android `AuthRepository`. Tokens live in `SessionController`.
/// `deviceJson` is produced only by a bound engine.
public final class AuthRepository: @unchecked Sendable {
    public let client: OlloClient
    private let sessions: SessionController?

    public init(client: OlloClient, sessions: SessionController? = nil) {
        self.client = client
        self.sessions = sessions
    }

    public func requestOtp(phone: String) async throws -> (challengeId: String, devOtp: String?) {
        let raw = try await client.requestOTP(phone: phone)
        return try AuthPayload.parseOtpChallenge(raw)
    }

    public func verify(
        challengeId: String,
        otp: String,
        accountEd25519: String,
        deviceJson: String,
        registrationLock: String? = nil
    ) async throws -> AuthSession {
        let body = try AuthPayload.verifyBody(
            challengeId: challengeId,
            otp: otp,
            accountEd25519: accountEd25519,
            deviceJson: deviceJson,
            registrationLock: registrationLock
        )
        let raw = try await client.verifyOTP(body: body)
        let session = try AuthPayload.parseSession(raw)
        try sessions?.save(
            SessionSecrets(
                userId: session.userId,
                deviceId: session.deviceId,
                access: session.access,
                refresh: session.refresh
            )
        )
        return session
    }

    /// Fail closed before burning an OTP: unbound engines throw here.
    public func signIn(
        engine: CryptoEngine,
        account: AccountKey,
        phone: String,
        otp: String,
        name: String,
        platform: String,
        registrationLock: String? = nil
    ) async throws -> AuthSession {
        let deviceJson = try engine.deviceRegistrationJson(name: name, platform: platform)
        let challenge = try await requestOtp(phone: phone)
        return try await verify(
            challengeId: challenge.challengeId,
            otp: otp,
            accountEd25519: account.publicB64(),
            deviceJson: deviceJson,
            registrationLock: registrationLock
        )
    }

    /// Primary registration: prove possession of the account Ed25519 key.
    public func signInWithKey(
        engine: CryptoEngine,
        account: AccountKey,
        name: String,
        platform: String,
        registrationLock: String? = nil
    ) async throws -> AuthSession {
        let deviceJson = try engine.deviceRegistrationJson(name: name, platform: platform)
        let raw = try await client.authChallenge()
        let challenge = try AuthPayload.parseAuthChallenge(raw)
        let proof = IdentityAddress.authProof(challengeId: challenge.challengeId, nonce: challenge.nonce)
        let signature = try account.sign(message: proof).base64EncodedString()
        let body = try AuthPayload.registerKeyBody(
            challengeId: challenge.challengeId,
            accountEd25519: account.publicB64(),
            signature: signature,
            deviceJson: deviceJson,
            registrationLock: registrationLock
        )
        let sessionRaw = try await client.registerKey(body: body)
        let session = try AuthPayload.parseSession(sessionRaw)
        try sessions?.save(
            SessionSecrets(
                userId: session.userId,
                deviceId: session.deviceId,
                access: session.access,
                refresh: session.refresh
            )
        )
        return session
    }

    public func logout() {
        sessions?.wipe()
    }

    /// Authenticated client: 401 retries once via refresh, then wipes the
    /// protocol store. [deviceJson] is still produced only by a bound engine.
    public static func connected(
        baseURL: URL,
        sessions: SessionController,
        urlSession: URLSession = .shared
    ) -> AuthRepository {
        let holder = ClientHolder()
        let client = OlloClient(
            baseURL: baseURL,
            session: urlSession,
            token: { sessions.access() },
            refreshTokens: {
                guard let token = sessions.refresh() else { return false }
                guard let client = holder.client else { return false }
                do {
                    let raw = try await client.refreshSession(token)
                    let pair = try AuthPayload.parseRefresh(raw)
                    return try sessions.applyRefresh(access: pair.access, refresh: pair.refresh)
                } catch {
                    return false
                }
            },
            onWipe: { sessions.wipe() }
        )
        holder.client = client
        return AuthRepository(client: client, sessions: sessions)
    }
}

private final class ClientHolder: @unchecked Sendable {
    var client: OlloClient?
}
